Move file attribute setting in WebsiteDataStore::resolveDirectoriesIfNecessary off main thread

szewai opened the pull request:
#### cee0c766f33c0b8ff0483f0b96d925150c740a79
<pre>
Move file attribute setting in WebsiteDataStore::resolveDirectoriesIfNecessary off main thread
<a href="https://bugs.webkit.org/show_bug.cgi?id=250821">https://bugs.webkit.org/show_bug.cgi?id=250821</a>
rdar://103883105

Reviewed by Youenn Fablet.

Setting excluded attribute might require disk access, so moving it to background thread.

* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::resolveDirectoriesIfNecessary):

Canonical link: <a href="https://commits.webkit.org/259121@main">https://commits.webkit.org/259121@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3deb88ae4f80491b33bfa22d9173685a805325c7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103940 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13057 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/36882 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/113171 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/14085 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3951 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/96190 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112263 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109712 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/10860 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/93928 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/38568 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92698 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25539 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80220 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6414 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/26920 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6575 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/3455 "Found 1 new test failure: fast/images/avif-heif-container-as-image.html (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12566 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46450 "Passed tests") | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/74/builds/6280 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8349 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -342,18 +342,23 @@ void WebsiteDataStore::resolveDirectoriesIfNecessary()
+            , resolvedNetworkCacheDirectory()
+    #if ENABLE(ARKIT_INLINE_PREVIEW)
+            , resolvedModelElementCacheDirectory()
+    #endif
```
Spaces here and above.

szewai replied:
will remove..

youennf commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -342,18 +342,23 @@ void WebsiteDataStore::resolveDirectoriesIfNecessary()
-    };
+    // These cache paths are not under Caches directory, so we need to explicitly
+    // exclude them from backup.
+    if (m_configuration->identifier()) {
```
Why this new check?
Also, we might be creating directories for some of the configuration resolved path, is it something we should also do in a background thread?

szewai replied:
For this new check: for WebsiteDataStores created without identifier, the default paths of these types are under Caches/ or tmp/ directory, which are excluded by the system; for WebsiteDataStores created with identifier, default paths of all types are under WebsiteDataStore/[identifier]/ directory, so we need to excluded them explicitly.
Therefore we only need to exclude forWebsiteDataStores created with identifier.

I remember it's because sandbox extension needs to be created on existing paths, and we need to create sandbox extension for parameters.

youennf replied:
Can you add this comment in thé commit message?

szewai committed:
Move file attribute setting in WebsiteDataStore::resolveDirectoriesIfNecessary off main thread
https://bugs.webkit.org/show_bug.cgi?id=250821
rdar://103883105

Reviewed by Youenn Fablet.

Setting excluded attribute might require disk access, so moving it to background thread.

* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::resolveDirectoriesIfNecessary):

Canonical link: https://commits.webkit.org/259121@main
